rfcs/0032-blockchain-in-hard-fork

```diff
@@ -0,0 +1,153 @@
+    maintain chain strength
+
+3. Retain the original genesis timestamp, use a time offset for new blocks
+
+## Detailed description of alternatives
+[detailed-descriptions]: #detailed-descriptions
+
+Alternative 1
+=============
+
+- choose a root and ledger at the pause point
+
+- the initial epoch, slot are the special case of 0,0
+   - otherwise, increment the slot from the pause point
+   - provide as a compile-time parameter
+
+- if we have a proof of the protocol state, provide that
+
+- subsequent blocks may use different proof, verification keys
+
+- force a software upgrade by bumping the protocol major version
+
+Notes
+-----
+
+- nodes will bootstrap from the new genesis state
+
+- time-locked accounts and transactions with expiration should continue
+   to work, because they rely on global slot numbers, and the
+   slot is incremented from the pause point
+
+Issues
+------
+
+- what is the source of randomness (epoch seed) for the post-fork chain?
+
+- is it acceptable to erase history (but we already have a succinct blockchain)
+
+- specifically, is there unfairness to the chains discarded past the pause point
+
+- what is allowed to change after the fork, with what effect:
+   - protocol state structure
+   - ledger structure/depth,
+   - account structure
+   - consensus constants
+
+- do we have to maintain pre-, post-fork code/data structures to verify the new genesis state,
+   ledger, and new blocks
+
+- if there's no proof of the ledger, how to convince the world that the new genesis ledger and
+   ledger are valid
+
+Alternative 2
+=============
+
+- choose a root and ledger at a pause point
+
+- nodes have a notion of "last fork time", an epoch-and-slot
+
+- upon a fork, gossip a special block to update the last fork time is issued, also referring to the
+   protocol state and proof at the pause point, and resetting the chain strength
+
+- if items affecting keys, ledger hashes, and proofs haven't changed, a software upgrade may not be required;
+   otherwise, force an upgrade
+
+Issues
+------
+
+- there may be empty slots or even empty epochs between the pause point and the restart; the reset
+   of the chain strength may be sufficient to handle the consequences
+
+- could we provide a new genesis block and proof, instead of the special block
+
+- As in Alternative (1), items may change after the fork, so there may be pre-, post-fork proofs, coda, data structures;
+   each additional fork may introduce more complexity
+
+- If the ledger/account structure changes, we won't have a proof for the post-fork state, also mentioned as
+   an issue for Alternative (1)
+
+- If the epoch, slot widths change, computations for epochs or slots may become more complex; subsequent forks may introduce
+   more complexity
+    - possible solution: a most-recent-fork timestamp (an earthly time, not the "last fork time") might be
+       used for epoch and slot calculations
```

Comment:
For epoch and slot calculations, dont we need both the pause time t_p and the restart time (t_r)? I mean, let f: t -> s be the function that took the earthly time as input and output the slot number. The new function f' could work as:
if t < t_p, then output f(t)
if t > t_r, then output f(t-gap), where gap = t_r - t_p

Reply:
yes, you'd have to treat times before the pause differently. If slot or epoch widths change after the restart, you might also need different functions.